Allow configuration in Main

nishantjr opened the pull request:
Heya!

I've made a PR that better separates personal configuration from the main code. This is a first step towards allowing different developers to have different configurations by importing this project as a library. It also makes a small effort to make writing configurations a little less intimidating by abstracting away FFI.

This is meant to be a quick and dirty PR and doesn't deal with all problems. For example, the Config.hs file should probably be renamed, but I wanted to gauge interest in this direction before making more extensive changes.

nishantjr committed:
Structure Main.hs more like XMonad's configuration file

We move the main-loop implementation to Config.hs, and call it with appConfig

nishantjr committed:
Use modKey configuration parameter

nishantjr committed:
Extract type State for encapsulating run-time state

nishantjr committed:
Move customKeybindings to Main

l-Shane-l commented:
Hey, I like the idea I am very conscious of how this code base might be for people with fresh eyes.

But im not sure about putting the configuration into Main.hs, the config will only grow and with custom keybindings and startup apps it will likely become long and verbose.  

Perhaps the Config.hs you propose could be renamed and Config.hs used as just configuration with none of the ffi complexity you mentioned.

The config or what ever we decide to call it will only grow in the near future as more features are added.

Do you think it would be better to have a default configuration and a custom configuration so a user only really needs to concern themselves with what they are interested in.

With Main I wanted this to manage the flow of the compositor from start to end, perhaps this is unnecessary or confusing? Maybe I should move that elsewhere and keep Main simple?

With regard interest, definitely yes, im very aware of how complex the code is becoming and I really want to make this usable. I think you are in a much better position than me to inform me on this aspect.

So in short what do you think about a further refactor?

- Simplifying Main.hs
- Providing a custom config file
- Potentially rename or even further dividing Config.hs perhaps changing it to something that support the custom config with default options

nishantjr commented:
Hi Shane, sorry about not responding earlier. Things have gotten a bit busy for me and I'm not sure when I can get back to it.
I agree with most of your suggestions---in particular Config.hs is not a great name.

One thing I feel strongly about is that the user configuration (i.e. what is in Main.hs in this PR), and cannot be depended on by other code, such as the main loop. This is so that this repository can be used as a library where users/developers provide their own custom configuration.

l-Shane-l commented:
Hey glad to see your reply tbh. Looking over your suggested changes it makes sense to me now. Your goals of making the configuration more modular and abstracting away FFI complexity make a lot of sense. I can see how this will make it easier for developers to create their own configurations when using this as a library.

I added a lot of features over xmass and pushed this code, It looks like it created a conflict. I think its just the starting apps section in config.hs, this likely would go in Main.hs in your approach?

Would you be able to resolve this conflicts? Once that's done, I'm happy to merge this and you can rename Config.hs if you think that would better reflect its purpose, perhaps compositorConfig or something like that.

I will hold off on any more code merges until you have time to fix the conflict. With the changes I made over xmass there are some more options to put in the Main config, like border color, thickness etc.

So in short if you can address the conflict I will merge this, if you dont have the time or cant resolve the conflict I can do this and merge the code.


Thanks for doing this btw, its not even something that crossed my mind. If you have any more suggestions id love to hear them.

nishantjr committed:
Extract actions: cycleWindows and terminate

nishantjr committed:
Config: Add onStartup param, and use that to implement startingApps

nishantjr commented:
Hi Shane. Glad we're on the same page :-) 

Looks like you've made great progress, nice!

I've rebased and resolved conflicts, but I've held off on the rename... I was wondering if we want to combine Config.hs and Setup.hs into a "TinyWLHS.hs" for now and split things out when we have a better idea of what the structure should be?

l-Shane-l commented:
Looks great, Ill merge it now, thank you.

Yes maybe a TinyWLHS.hs or maybe move in into the library and keep the app as a reference app. I have been thinking about this a lot actually. The project still have not fully found its identity and is trying to be too many things. It was originally just suppose to test bindings. 

I think it should be just a library for creating wayland compositors in Haskell. The app then should just be a reference implementation, so there is probably space for a dedicated Wayland Haskell compositor focused solely on that.

nishantjr commented:
> It was originally just suppose to test bindings.

Oh! Didnt't realise that.
May I ask if you attempted to get the existing [waymonad] project going. A lot of work seems to be done there... I wonder if it would make sense putting in the effort to revive it?

[waymonad]: https://github.com/waymonad/waymonad

l-Shane-l commented:
I did and its various forks and there are other attempts at a haskell wayland compositor. If you have the time you can read through this thread https://discourse.haskell.org/t/xmonad-for-wayland-call-for-help/7812/1 and then this one https://discourse.haskell.org/t/haskell-wlroots-bindings/8426

I appear at the end of the second. So there have been numerous attempts and all seem to have been abandoned, with the most recent stuck on a PR. The core issues seemed to be the scope of work but equality confusion around wayland and its various parts.

So tiny-wlhs here based on the tinywl reference reference implementation in wlroots. The idea was to start with something minimal and that could be tested rather than trying to get blanket bindings for wlroots. When I started the project I intended to use the existing haskell wlroots bindings and to provide a basic app to test them and provide some feedback. People in the threads were complaining about not knowing who the bindings were for or what the purpose of the work was.

When I tried to start using the bindings there was not even a working binding to logging. I added this and created a PR to wlhs but this is still sitting there. 

One of the main challenges of wayland is that is an evolving protocol defined in xml, this needs to be converted to code and its heavily opinionated towards C like language. So in C you run wayland scanner to generate the code you need. In haskell that would mean either generating the C code from XML and then binding (not ideal as most bindings need to be hand written) or create a haskell scanner for the xml. I briefly investigated the later. If I had more time I would do the scanner. 

Instead though I thought it would just be easier to take tinywl with all the wayland, wayland scanner and wlroots complexity hidden behind a single c file and its make file. I took that file and converted it to a shared library and bound that instead. In the future as wayland evolves and changes I can be sure that the tinywl.c in wlroots will be maintained and I just need to stay up to date with that. 

Still though this repo is set up so when/if any haskell bindings to wlroots are created they can be used and test here. To me at least it made things a lot simple to focus on a compositor implementation and add to it piece by piece with a focus on features. The short term goal was/is to get something I can use as a daily driver and the assumption is from there It can be improved interactivity. Also it just seemed the best way to learn about wayland and what was needed for a haskell implementation.

So that is how it evolved. Originally I thought I would just convert everything to Haskell but I think after working in it for a few months leaving the complexity in C where there is so much documentation and reference material makes more sense. The protocol is very opinionated towards C and C is just a lot more perform-ant. The questions is more about what to expose in Haskell, for now I think any control or configure logic should be in Haskell.

Im very conscious of the previous abandoned efforts so with this I put a premium on project sustainability and keeping the maintenance cost down.

With regard the projects future I want to get to that daily driver phase and I think for me that would be with the Copy/Paste functionality added. After that tiling and multi monitor would come next.

In the meantime I want to add some github action builds and start on tests, I think better documentation is needed likely on its own site (perhaps on github) and I plan to start a formal dev log. I will add all this and the future features into issues, in case anyone else is interested in implementing them.

nishantjr commented:
Oh, that makes a lot of sense. Sounds like a solid plan. I'd be happy to put in some time here and there... especially if you've got concrete tasks that need working on.
